src/python/lib/ensembl/compara/citest/pytest_citest.py

```diff
@@ -0,0 +1,159 @@
+
+import py
+import pytest
+from _pytest.config.argparsing import Parser
+from _pytest.runner import TestReport
+
+from ensembl.compara.db.dbconnection import DBConnection
+from ensembl.compara.filecmp.dircmp import DirCmp
+from ensembl.compara.citest import testdb, testfiles
+
+
+@pytest.hookimpl()
+def pytest_addoption(parser: Parser) -> None:
+    """Register argparse-style options for CITest."""
+    group = parser.getgroup("continuous integration test (citest)")
+    group.addoption('--ref-db', action='store', metavar='URL', dest='ref_db',
+                    help="URL to the reference database")
+    group.addoption('--ref-dir', action='store', metavar='PATH', dest='ref_dir',
+                    help="Path to reference's root directory")
+    group.addoption('--target-db', action='store', metavar='URL', dest='target_db',
+                    help="URL to the target database")
+    group.addoption('--target-dir', action='store', metavar='PATH', dest='target_dir',
+                    help="Path to target's root directory")
+
+
+def pytest_collect_file(parent: pytest.Session, path: py.path.local) -> Optional[pytest.File]:
+    """Returns the collection of tests to run as indicated in the given JSON file."""
+    if path.ext == '.json':
+        return JsonFile(path, parent)
+    return None
+
+
+def pytest_sessionstart(session: pytest.Session) -> None:
+    """Add required variables to the session before entering the run test loop."""
+    session.report = {}
+
+
+@pytest.hookimpl(tryfirst=True, hookwrapper=True)
+def pytest_runtest_makereport(item: pytest.Item) -> TestReport:
+    """Returns the test report updated with custom information."""
+    outcome = yield
+    report = outcome.get_result()
+    if report.when == 'call':
+        item.session.report[item] = report
+
+
+def pytest_sessionfinish(session: pytest.Session) -> None:
+    """Generate a custom report before returning the exit status to the system."""
+    # Use the configuration JSON file as template for the report
+    config_filename = session.config.getoption('file_or_dir')[0]
+    with open(config_filename) as f:
+        full_report = json.load(f, object_pairs_hook=OrderedDict)
+    # Update/add global information
+    full_report['reference_db'] = session.config.getoption('ref_db', full_report.get('reference_db'), True)
+    full_report['reference_dir'] = session.config.getoption('ref_dir', full_report.get('reference_dir'), True)
+    full_report['target_db'] = session.config.getoption('target_db', full_report.get('target_db'), True)
+    full_report['target_dir'] = session.config.getoption('target_dir', full_report.get('target_dir'), True)
+    # Add the reported information of each test
+    failed = 0
+    for item, report in session.report.items():
+        if isinstance(item, testdb.TestDBItem):
+            test_list = full_report['database_tests'][item.table]
+        else:
+            test_list = full_report['files_tests']
```

Comment:
This and the `JsonFile` class below know maybe a bit too much about sub-classes. It could maybe be improved with be some sort of dispatcher, virtual methods, etc. *But* I don't think it's worth refactoring :relieved:. This module is clear enough and straightforward.